[Draft] Use native file handlers

marecl opened the pull request:
This PR changes how IOFile is handled on OS-level.  
Improves compatibility and granularity with Orbis file-related kernel calls.  
Existing (in-app call) behaviour and usage is not affected (I think).  

Currently implemented for UNIX (POSIX calls).  
Windows implementation pending.  
Further tests are needed to confirm correct operation.  

Subject to changes.

marecl committed:
Moved file handling to native calls:
* Removed streams, files are handled by native file descriptor (Linux only)
* Removed FileType - unnecessary, native handlers don't divide text/binary
* Divided io_file.cpp between linux/posix and windows, selected upon compiling (reduce mess in source files)
* Files canm be accessed without changes (FileAccessMode) - target behaviour preserved, added option to truncate files on writing (Write/WriteExtended)
* Minor commenting and documentation
* Checks, modified flow of open() in file_system.cpp

marecl committed:
Moved file handling to native calls:
* Removed streams, files are handled by native file descriptor (Linux only)
* Removed FileType - unnecessary, native handlers don't divide text/binary
* Divided io_file.cpp between linux/posix and windows, selected upon compiling (reduce mess in source files)
* Files canm be accessed without changes (FileAccessMode) - target behaviour preserved, added option to truncate files on writing (Write/WriteExtended)
* Minor commenting and documentation
* Checks, modified flow of open() in file_system.cpp

marecl committed:
Disambiguation, moving parameter resolving to target OS source

marecl committed:
Disambiguation, moving parameter resolving to target OS source

marecl committed:
how did these get here

marecl committed:
how did these get here

marecl committed:
errno (and Windows equivalent) propagation

marecl committed:
errno (and Windows equivalent) propagation

marecl committed:
clang

marecl committed:
clang

marecl committed:
Merge branch 'main' of https://github.com/shadps4-emu/shadPS4 into posixiofile

marecl committed:
Merge branch 'posixiofile' of https://github.com/marecl/shadPS4 into posixiofile

marecl committed:
Adjusted naming convention (flags [open], mode [permissions])
File permissions are passed on posix_open

marecl committed:
IOFile is platform-agnostic (again)
Platform-specific binds moved to separate namespace

marecl committed:
Implemented Exists()

marecl committed:
Implemented CreateDirectory, CreateDirectories

marecl committed:
De-duplicated routines, throwables call noexcept overloads

marecl committed:
C binds are now EXCEPTIONal huehuehue
Doesn't compile, I need a break

marecl committed:
Mostly implemented

marecl committed:
memo

marecl committed:
Fix sdl version not compiling

marecl committed:
clang

marecl committed:
gcc or mac, idk

marecl committed:
Cleanup

marecl committed:
stuff

marecl committed:
doesn't write stdout to random files

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -108,51 +107,77 @@ s32 PS4_SYSV_ABI open(const char* raw_path, s32 flags, u16 mode) {
+            Kernel_Fs,
+            "Read and Truncate flags specified. This makes no sense, neutering Truncate flag");
+        truncate = false;
+    }
```
We're emulating the PS4, not the POSIX spec.
Undefined behavior or not, we should follow how a PS4 behaves.
Since the truncate flag causes an error return in a read-only file system, and we emulate that properly, any data loss visible in emulator would be visible on real hardware too.

marecl replied:
Change: File is truncated in this check, flag is removed, opened normally later in the function

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -108,51 +107,77 @@ s32 PS4_SYSV_ABI open(const char* raw_path, s32 flags, u16 mode) {
+    }
 
-        file->type = Core::FileSys::FileType::Directory;
+    if (directory && (write || rdwr || truncate)) {
```
More of a nitpick than anything, but modifying this check to include truncate is unnecessary since truncate with any form of write access hits an earlier error return.

marecl replied:
Change: removed `truncate`

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -162,63 +187,17 @@ s32 PS4_SYSV_ABI open(const char* raw_path, s32 flags, u16 mode) {
+        e = file->f.Open(file->m_host_name, Common::FS::AccessModeOrbisToPOSIX(flags), mode);
     }
 
     if (e != 0) {
```
From what I can see, your `f.Open` call returns the native result (-1 on fail).
If you want to keep that as-is, this check here needs to pass through and check an std::error_code for the error instead.
Alternatively, switch `f.Open` to return the `std::error_code` value like it does on main, and this code will function properly.

marecl replied:
No change

marecl committed:
stuff

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -1041,15 +1015,16 @@ s32 PS4_SYSV_ABI posix_unlink(const char* path) {
         // File to unlink hasn't been opened, manually open and unlink it.
-        Common::FS::IOFile file(host_path, Common::FS::FileAccessMode::ReadWrite);
+        // marecl: *may* fail. IDK if it's supposed to exist (yet)
+        Common::FS::IOFile file(host_path, Common::FS::FileAccessMode::ReadExtended);
```
If the file to unlink doesn't exist, we should be setting errno to ENOENT and returning -1.
I believe that case should be caught by the `mnt->GetHostPath` call above though.

marecl replied:
Change: removed comment

StevenMiller123 commented on src/core/libraries/kernel/file_system.cpp:
```diff
@@ -108,51 +107,77 @@ s32 PS4_SYSV_ABI open(const char* raw_path, s32 flags, u16 mode) {
+        h->DeleteHandle(handle);
+        *__Error() = POSIX_EINVAL;
+        return -1;
+    }
```
Also more of a nitpick, but this will never be hit.
read will always be true by this point, since if neither read, write, or rdwr is specified, then the call will hit the EINVAL error case at the start of this function, and if either write or rdwr is specified, the call hits the EISDIR error case instead.

marecl replied:
Change: Removed block

StevenMiller123 reviewed (changes requested):
I've looked primarily at the changes to file_system.cpp in this review, since I wrote much of the code currently in there.
I haven't dug too deeply into your changes to flag handling in `open`, but I know there's some incorrect changes there too.

marecl committed:
minor cleanup

marecl committed:
reviewed changes

marecl committed:
Cleanup cont'd
Corrected namespaces
IOFile IsOpen quits earlier if fd==-1
Removed misleading/redundant comments

marecl committed:
Merge branch 'main' into posixiofile

marecl committed:
Resolve compilation errors after merge from main

marecl committed:
Revert accidental mute, clang

marecl committed:
Merge branch 'main' into posixiofile

marecl committed:
Corrected behaviour to match PS4's
Added NativeFS::Touch()
Translated all POSIX errors to Orbis

marecl committed:
clang

marecl committed:
Catch errors from read/write
R/W returning s64 (signed), not size_t (unsigned)

marecl committed:
Merge remote-tracking branch 'upstream/main' into posixiofile
